pkg/controller/podgc/gc_controller_test.go

```diff
@@ -135,90 +150,203 @@ func TestGCTerminated(t *testing.T) {
-		}
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			nodeList := &v1.NodeList{}
+			for _, nodeName := range test.initialClientNodes {
+				nodeList.Items = append(nodeList.Items, *testutil.NewNode(nodeName))
+			}
+			client := fake.NewSimpleClientset(nodeList)
+			gcc, podInformer, nodeInformer := NewFromClient(client, -1)
+			fakeClock := clock.NewFakeClock(time.Now())
+			gcc.nodeQueue = NewBasicUniqueWorkQueue(fakeClock)
 
-		creationTime := time.Unix(0, 0)
-		for _, pod := range test.pods {
-			creationTime = creationTime.Add(1 * time.Hour)
-			podInformer.Informer().GetStore().Add(&v1.Pod{
-				ObjectMeta: metav1.ObjectMeta{Name: pod.name, CreationTimestamp: metav1.Time{Time: creationTime}},
-				Status:     v1.PodStatus{Phase: pod.phase},
-				Spec:       v1.PodSpec{NodeName: "node"},
-			})
-		}
+			deletedPodNames := make([]string, 0)
+			var lock sync.Mutex
+			gcc.deletePod = func(_, name string) error {
+				lock.Lock()
+				defer lock.Unlock()
+				deletedPodNames = append(deletedPodNames, name)
+				return nil
+			}
 
-		pods, err := podInformer.Lister().List(labels.Everything())
-		if err != nil {
-			t.Errorf("Error while listing all Pods: %v", err)
-			return
-		}
-		gcc.gcOrphaned(pods)
+			for _, nodeName := range test.initialListerNodes {
+				nodeInformer.Informer().GetStore().Add(testutil.NewNode(nodeName))
+			}
+			creationTime := time.Unix(0, 0)
+			for _, pod := range test.pods {
+				creationTime = creationTime.Add(1 * time.Hour)
+				podInformer.Informer().GetStore().Add(&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{Name: pod.name, CreationTimestamp: metav1.Time{Time: creationTime}},
+					Status:     v1.PodStatus{Phase: pod.phase},
+					Spec:       v1.PodSpec{NodeName: pod.nodeName},
+				})
+			}
 
-		pass := true
-		for _, pod := range deletedPodNames {
-			if !test.deletedPodNames.Has(pod) {
-				pass = false
+			// First GC of orphaned pods
+			gcc.gc()
+			if len(deletedPodNames) > 0 {
+				t.Errorf("no pods should be deleted at this point.\n\tactual: %v", deletedPodNames)
 			}
-		}
-		if len(deletedPodNames) != len(test.deletedPodNames) {
-			pass = false
-		}
-		if !pass {
-			t.Errorf("[%v]pod's deleted expected and actual did not match.\n\texpected: %v\n\tactual: %v", i, test.deletedPodNames, deletedPodNames)
-		}
+			// Move clock forward
+			fakeClock.Step(test.delay)
+
+			// Execute planned nodes changes
+			for _, event := range test.nodeWatchEvents {
+				node := testutil.NewNode(event.name)
+				switch event.kind {
+				case watch.Added:
+					nodeInformer.Informer().GetStore().Add(node)
+				case watch.Deleted:
+					nodeInformer.Informer().GetStore().Delete(node)
+				}
+			}
+			for _, change := range test.clientChanges {
+				node := testutil.NewNode(change.name)
+				switch change.kind {
+				case watch.Added:
+					client.CoreV1().Nodes().Create(node)
+				case watch.Deleted:
+					client.CoreV1().Nodes().Delete(change.name, &metav1.DeleteOptions{})
+				}
```

Comment:
Yeah - the more I look into this test, the more I'm convinced we should modify informers.
I think you should just update client, then List from client and pass the list of pods and nodes to gcOrphaned.

Reply:
I am explicitly testing scenarios when client and informers are not in sync